### PR TITLE
474406 feature/gnav  2020

### PR DESCRIPTION
**Pull request by maccesar21**
## Global Navigation
Reworks structure of global navigation, and styling for multiple viewports.

### CODE REVIEW

File: src/header/js/header-menus.js [hunk 1]
```diff
@@ -1,5 +1,6 @@
 import {$assign as $, $dispatch, $replaceAll, $renderSvgOrImg} from '../../shared/js/shared';
 import {$hamburger} from '../../shared/js/iconPaths';
+import Hammer from 'hammerjs';
```
Comment (paulcpederson) [inline src/header/js/header-menus.js]: are you using this at all? I don't see any references to this API...

File: src/header/js/header-menus.js [hunk 2]
```diff
@@ -307,6 +313,135 @@ export default ({variant = 'desktop'}) => {
...
+		const active = categoryList.getAttribute('aria-expanded') === 'false' ? 'true' : 'false';
+		const categoryDetailsItems = [].slice.call(category.target.parentNode.querySelectorAll('.esri-header-menus-flyout--categories-details_item'));
+		const catsComputedStyle = window.getComputedStyle(categoryDetailsItems[0]);
+		const computedHeight = (parseInt(catsComputedStyle.height) * categoryDetailsItems.length);
+		const computedMargin = (parseInt(catsComputedStyle.marginTop) * categoryDetailsItems.length) + parseInt(catsComputedStyle.marginTop);
+		const headers = [].slice.call(document.querySelectorAll('.esri-header-menus-flyout--categories-item_header'));
+		const items = [].slice.call(document.querySelectorAll('.esri-header-menus-flyout--categories-item'));
+		const itemsList = [].slice.call(document.querySelectorAll('.esri-header-menus-flyout--list-items'));
+		const viewport = (window.innerWidth < 1024) && 'mobile';
+
```
Comment (dlewis-irse) [inline src/header/js/header-menus.js]: i would rename this to isMobile and remove the &&...

File: src/header/js/header-menus.js [hunk 3]
```diff
@@ -307,6 +313,135 @@ export default ({variant = 'desktop'}) => {
...
+				item.addEventListener('click', (e) => {
+					const parentNode = e.target.parentNode;
+					const selectedCategory = parentNode.getAttribute('data-id');
+
+					itemsList.forEach((list, index) => {
+						list.setAttribute('aria-current', 'false');
+						items[index].setAttribute('aria-current', 'false');
+					});
+
+					parentNode.setAttribute('aria-current', 'true');
```
Comment (dlewis-irse) [inline src/header/js/header-menus.js]: shouldn't you only set this when the menu is selected

Comment (maccesar21) [reply src/header/js/header-menus.js]: @dlewis-irse updated..

**Comment by dlewis-irse** [comment]
@maccesar21 can you please leave one menu authored as structured so we can see all 3 possible menu types in the example data file?

File: src/header/js/header-menus.js [hunk 4]
```diff
@@ -237,6 +244,68 @@ export default ({variant = 'desktop'}) => {
 		);
 	});
 
+	function resetFlyoutDimensions(parentNode) {
+		const subMenus = document.querySelectorAll('.esri-header-menus-submenu');
+		const parent = (parentNode !== 'disabled' && parentNode !== 'init') && parentNode.getAttribute('data-parent');
+		const parentElement = document.querySelector(`#${parent}`);
+
+		if (parentNode === 'init') {
```
Comment (dlewis-irse) [inline src/header/js/header-menus.js]: Is ParentNode a node or a string?  The name makes me think it should be a DOM element, but I see it's just a string in the if statments

File: src/header/js/header-menus.js [hunk 5]
```diff
@@ -237,6 +244,68 @@ export default ({variant = 'desktop'}) => {
...
+	function resetFlyoutState() {
+		const flyoutCategories = [].slice.call(document.querySelectorAll('.esri-header-menus-flyout--categories-item'));
+		const flyoutList = [].slice.call(document.querySelectorAll('.esri-header-menus-flyout--list-items'));
+
+		if (flyoutList.length) {
+			flyoutList.forEach((list, index) => {
+				flyoutCategories[index].setAttribute('aria-current', 'false');
+				list.setAttribute('aria-current', 'false');
+				if ((list.hasAttribute('data-id') && list.getAttribute('data-id') === '0') && 
+						(flyoutCategories[index].hasAttribute('data-id') && flyoutCategories[index].getAttribute('data-id') === '0')
```
Comment (dlewis-irse) [inline src/header/js/header-menus.js]: you don't need the hasAttribute checks here - just the getAttribute checks will do the same thing

File: src/header/js/header-menus.js [hunk 6]
```diff
@@ -307,6 +377,148 @@ export default ({variant = 'desktop'}) => {
...
+				categoryList.style.height = '0';
+				categoryHeader.setAttribute('aria-current', 'false');
+			}
+		} else {
+			items.forEach((item, index) => {
+				item.addEventListener('click', (e) => {
+					const parentNode = e.target.parentNode;
+					const selectedCategory = parentNode.hasAttribute('data-id') && parentNode.getAttribute('data-id');
+					const selectedList = itemsList[index].hasAttribute('data-id') && itemsList[index].getAttribute('data-id');
+					const selectedListCols = itemsList[index].hasAttribute('data-coltype') && itemsList[index].getAttribute('data-coltype');
```
Comment (dlewis-irse) [inline src/header/js/header-menus.js]: you don't need the hasAttribute checks here

File: src/header/js/header-menus.js [hunk 7]
```diff
@@ -307,6 +377,148 @@ export default ({variant = 'desktop'}) => {
...
+			});
+		}
+	}
+	
+	function renderFlyoutMenu(items, type, id, uuid, suuid) {
+		const $items = [];
+		const listArr = [];
+		let category = "";
+
+		if (type === 'category') {
```
Comment (dlewis-irse) [inline src/header/js/header-menus.js]: this should be a switch statement - not chained if statements since you evaluating the same variable

Comment (dlewis-irse) [reply src/header/js/header-menus.js]: the check of items.cols && items.cols.length is also repeated that should only need to be checked once

Comment (maccesar21) [reply src/header/js/header-menus.js]: @paulcpederson I think this is the final updates.. Can I please get your review on this?

**Comment by dlewis-irse** [comment]
@paulcpederson Cesar is now dev complete - do you see anything you want changed before we merge this?  (I will be doing a Squash and Merge)

**Comment by paulcpederson** [comment]
👀 looking now

File: src/index.html [hunk 8]
```diff
@@ -7,6 +7,7 @@
     <link rel="stylesheet" href="./esri-global-nav.css">
     <link rel="stylesheet" href="https://s3-us-west-1.amazonaws.com/patterns.esri.com/files/calcite-web/1.0.1/css/calcite-web.min.css">
```
Comment (paulcpederson) [inline src/index.html]: if you're pulling in fonts from the CDN, you should use the no-fonts version of calcite-web (and also bump up the version as well)

File: src/header/css/header-menus.pcss [hunk 9]
```diff
@@ -235,7 +235,370 @@
...
+		display: table;
+		clear: both;
+	}
+
+	@media (--width-l) {
+		overflow: hidden;
+	}
+}
+
+.esri-header-menus-flyout--categories,
```
Comment (paulcpederson) [inline src/header/css/header-menus.pcss]: Are you trying to do BEM here? Usually sub-components inside a block are `__` while modifiers are `--`.

File: src/demo.js [hunk 10]
```diff
@@ -137,85 +137,242 @@ document.addEventListener("DOMContentLoaded", () => {
...
+								category: 'LOCATION-ENABLED PRODUCTS',
+								cols: [
+									{
+										col: [
+											{
+												heading: 'CAPABILITIES',
```
Comment (paulcpederson) [inline src/demo.js]: These headings should not be uppercase. If we wish to display them uppercase we should use `text-transform: uppercase;`. The reason is that some words when written in all uppercase are read as acronyms by screen readers.  Like the word "It" will be read as "I.T.".

**Comment by paulcpederson** [comment]
Screen Shot 2019-10-10 at 10 44 31 AM

- You are putting `li` elements inside of a `div` instead of a `ul`. This is invalid html
- Using a `p` for something that appears to be a heading. Should probably be a `h` or `role=heading`

- The `esri-header-menus-flyout--list-items_label` class is not accessible contrast on white

File: src/header/css/header-menus.pcss (hunk 9 above)
Comment (maccesar21) [reply src/header/css/header-menus.pcss]: I am not trying to do BEM.. This is just my preference.. I don't like to use _ unless it's a two_word or something..

**Comment by paulcpederson** [comment]
Cool, I think the `esri-header-menus-flyout--list-items_label` class is still too light, but we can fix in another pr. @dlewis-irse @maccesar21  do you two want to squash and merge this?

**Pull request merged**